libs/wingc/src/visit.rs

```diff
@@ -0,0 +1,270 @@
+use crate::ast::{
+	ArgList, Class, Constructor, Expr, ExprKind, FunctionDefinition, InterpolatedStringPart, Literal, Reference, Scope,
+	Stmt, StmtKind,
+};
+
+/// Visitor pattern inspired by implementation from https://docs.rs/syn/latest/syn/visit/index.html
+///
+/// If you are implementing a visitor, you can delegate back to the
+/// default implementation to visit the children of a node by calling
+/// visit::visit_<node_type>(self, node);
+///
+/// For example:
+///
+/// impl<'ast> Visit<'ast> for ExprVisitor {
+///   fn visit_item_fn(&mut self, exp: &'ast Expr) {
+///     println!("Expr with span={}", exp.span);
+///
+///     // Delegate to the default impl to visit any nested expressions.
+///     visit::visit_expr(self, exp);
+///   }
+/// }
+pub trait Visit<'ast> {
+	fn visit_scope(&mut self, e: &'ast Scope) {
+		visit_scope(self, e);
+	}
+	fn visit_stmt(&mut self, e: &'ast Stmt) {
+		visit_stmt(self, e);
+	}
+	fn visit_class(&mut self, e: &'ast Class) {
+		visit_class(self, e);
+	}
+	fn visit_constructor(&mut self, e: &'ast Constructor) {
+		visit_constructor(self, e);
+	}
+	fn visit_expr(&mut self, e: &'ast Expr) {
+		visit_expr(self, e);
+	}
+	fn visit_literal(&mut self, e: &'ast Literal) {
+		visit_literal(self, e);
+	}
+	fn visit_reference(&mut self, e: &'ast Reference) {
+		visit_reference(self, e);
+	}
+	fn visit_function_definition(&mut self, e: &'ast FunctionDefinition) {
+		visit_function_definition(self, e);
+	}
+	fn visit_args(&mut self, e: &'ast ArgList) {
+		visit_args(self, e);
+	}
```

Comment:
We can add a lot more cases for visiting more specific types of nodes (e.g. `visit_expr_binary_op`), the only limitation is we don't have dedicated structs with this information because they're currently inlined into the `Expr` enum variants (like [here](https://github.com/winglang/wing/blob/b0e799f14fb4a864eb26007ae45cb418389c5649/libs/wingc/src/ast.rs#L304-L309)). I can create those structs but I left it out of this PR to avoid changing too much at once / making the code harder to review.